packages/core/src/tracing/instrumentation/frameworks/fastify.js

```diff
@@ -47,51 +50,36 @@ function instrument(build) {
   function overwrittenBuild() {
     const app = build.apply(this, arguments);
 
-    if (app.route) {
-      const originalRoute = app.route;
-      app.route = function shimmedRoute(opts) {
-        if (opts.handler) {
-          const originalHandler = opts.handler;
-          opts.handler = function shimmedHandler() {
-            annotateHttpEntrySpanWithPathTemplate(app, opts);
-            return originalHandler.apply(this, arguments);
-          };
-        }
-
-        let preHandler;
-        let preHandlerKey;
-        if (opts.preHandler) {
-          // In Fastify 2.x, the attribute is called preHandler.
-          preHandler = opts.preHandler;
-          preHandlerKey = 'preHandler';
-        } else if (opts.beforeHandler) {
-          // In Fastify 1.x, the attribute is called beforeHandler.
-          preHandler = opts.beforeHandler;
-          preHandlerKey = 'beforeHandler';
-        }
-
-        if (preHandler) {
-          if (typeof preHandler === 'function') {
-            opts[preHandlerKey] = function shimmedPreHandler() {
-              annotateHttpEntrySpanWithPathTemplate(app, opts);
-              return preHandler.apply(this, arguments);
-            };
-          } else if (Array.isArray(preHandler)) {
-            opts[preHandlerKey].unshift(function prependedBeforeHandler(request, reply, done) {
-              annotateHttpEntrySpanWithPathTemplate(app, opts);
-              done();
-            });
-          }
-        }
-
-        return originalRoute.apply(this, arguments);
-      };
+    // NOTE: all major versions support `addHook` - this is just a safe protection
+    if (!app.addHook) {
+      logger.warn('Instana was not able to instrument Fastify. The instrumention of http requests is still working.');
+      return app;
     }
 
+    app.addHook('onRequest', function onRequest(request, reply, done) {
+      try {
+        // NOTE: v1 uses _context https://github.com/fastify/fastify/blob/1.x/fastify.js#L276
+        //       v2/v3 uses context https://github.com/fastify/fastify/blob/2.x/test/handler-context.test.js#L41
+        const url = reply._context ? reply._context.config.url : reply.context.config.url;
```

Comment:
That's why I have put a try/catch around the whole execution to avoid this.

> Both would be a pre-caution against breaking apps with future versions of fastify

If that happens it would log a warning and would not interrupt the customer.

I have not understood what the benefit is of changing the checks :)

Reply:
You mean fastify v4 would introduce `reply._context` and it would no longer work?

Reply:
The try/catch also works, I failed to realize that this is wrapped in a try/catch 🙈 

It should be an empty catch block though or at most log at debug level. If there is a problem with this, we would log a warning for every incoming request, that is far too noisy.

My suggestion for adding additional checks is due to the general rule that instrumentation code must never make any assumptions about the structure of incoming parameters etc., because these can change between versions of the instrumented framework. Even between patch versions, when we access internal of the instrumented framework that are not part of the public API.

So, ignoring the try/catch (which I didn't see) for a moment, my idea was to make the code here safe against `Cannot read properties of undefined` type of errors:  

* The current code checks if `reply._context` exists. If it exists, `reply._context.config.url` is read. Now, if `reply._context` exists, but `reply._context.config` does not exist, this code would produce an error.
* If `reply._context` does not exist, `reply.context.config.url` is accessed. If either `reply.context` does not exist  or `reply.context.config` does not exist, this code would produce an error.

Does that make sense?

Reply:
Yeah it makes totally sense 👍  And I like that we speak about it :) 

I decided against these specific checks and just went with try/catch because it's personally speaking "an enough protection" and it makes the code much more readable. 

Even if fastify changes the structure in any upcoming version (I doubt for v1, neither for v2), the http span is still captured, nothing breaks and the path template would be missing. I think this risk is okay to take?

I can only think of two use cases:

v3 or v4 adds `reply._context` and it won't find the url anymore.

Reply:
(: TBH I don't buy the argument of "only one use case" because that is making assumptions about the shape of `reply`, but anyway, the try-catch is good enough. 👍